---
title: "Plotting simulated PCR results"
bibliography: '`r sharedbib::bib_path()`'
output:
  html_document:
    css: style.css
---

```{r setup, include=FALSE}
source('style.R')
```

## Prepare

```{r}
library(ape)
library(metacoder)
library(tibble)
library(stringr)
library(readr)
library(dplyr)
library(furrr)
library(Biostrings)
```

## Plot Felipe's analysis results

### Read in results

These results were generated by Felipe Albornoz using Geneious.
They consist of two FASTA files: the reference sequences used for the simulated PCR and the predicted amplicons.

```{r}
ref_seqs <- read.FASTA(file.path("raw_data", "felipes_sim_pcr_results", "all.ref.seqs.rps10.fasta"))
amp_seqs <- read.FASTA(file.path("raw_data", "felipes_sim_pcr_results", "PCR product.fasta"))
```

I will check that all names in `amp_seqs` are in `ref_seqs`

```{r}
if (! all(names(amp_seqs) %in% names(ref_seqs))) {
  stop("All names in `amp_seqs` must be in `ref_seqs`")
}
```

Save location of named seqs in reference

```{r}
amped_indexes <- match(names(amp_seqs), names(ref_seqs))
```


### Clean up headers

The taxonomic information in the FASTA headers need to be cleaned up to be useful for computational analysis and presentation.

```{r}
clean_headers <- function(headers) {
  # remove _F , _sp
  headers <- sub(headers, pattern = "_F$", replacement = "")
  headers <- sub(headers, pattern = "_sp$", replacement = "")
  
  # Remove existing classifcaitons and only use seq ID
  headers[grepl(headers, pattern = "_Eukaryota;")] <- gsub(headers[grepl(headers, pattern = "_Eukaryota;")],
                                                           pattern = "_Eukaryota;.+$", replacement = "")
  
  # Remove everything after species name
  matches <- stringr::str_match(headers, "^([a-zA-Z]+_[a-zA-Z]+).*$")[,2]
  headers[! is.na(matches)] <- matches[! is.na(matches)]
  
  headers <- sub(headers, pattern = "_sp$", replacement = "")
  headers <- sub(headers, pattern = "_x$", replacement = "")
  
  # Remove underscores
  # headers <- gsub(headers, pattern = "([a-zA-Z]{3,})_", replacement = "\\1 ")
  
  headers <- sub(headers, pattern = "Peronosppora", replacement = "Peronospora")
  headers <- sub(headers, pattern = "Scleroderma", replacement = "Sclerospora")
  
  return(headers)
}
cleaned_headers <- clean_headers(names(ref_seqs))
```

Some headers have species names, others have genbank ids. 
I will read in seq ids and names separately and combine classifications.
Since looking up the taxonomy data can be unreliable depending on internet connection, I will cache the results of this in an .rds file.

Seq IDs:

```{r}
seq_id_obj_data_path <- file.path('intermediate_data', 'seq_id_obj.rds')
if (file.exists(seq_id_obj_data_path)) {
  seq_id_obj <- readRDS(seq_id_obj_data_path)
} else {
  seq_id_obj <- lookup_tax_data(cleaned_headers, type = "seq_id")
  saveRDS(seq_id_obj, file = seq_id_obj_data_path)
}
# seq_id_obj <- remove_redundant_names(seq_id_obj)
seq_id_classes <- classifications(seq_id_obj)[names(seq_id_obj$data$query_data)]
```

Seq names:

```{r}
genus_part <- sub(cleaned_headers, pattern = "^([a-zA-Z]+)_.*$",
                  replacement = "\\1")
species_part <- sub(cleaned_headers, pattern = "^[a-zA-Z]+_([a-zA-Z]+)$",
                  replacement = "\\1")
```

I will also save the results of this database search, since it can take a while and it requires manual input when synonyms are encountered, so cannot be included in an Rmd otherwise.

```{r}
taxon_name_obj_data_path <- file.path('intermediate_data', 'taxon_name_obj.rds')
if (file.exists(taxon_name_obj_data_path)) {
  taxon_name_obj <- readRDS(taxon_name_obj_data_path)
} else {
  taxon_name_obj <- lookup_tax_data(genus_part, type = "taxon_name")
  saveRDS(taxon_name_obj, file = taxon_name_obj_data_path)
}
taxon_name_classes <- classifications(taxon_name_obj)[names(taxon_name_obj$data$query_data)]
```

Combine the classifications from the name and ID searches:

```{r}
taxon_name_classes[taxon_name_classes != "unknown taxon"] <- paste0(taxon_name_classes, ";", species_part)[taxon_name_classes != "unknown taxon"]

combined_class <- unname(seq_id_classes)
combined_class[combined_class == "unknown taxon"] <- taxon_name_classes[combined_class == "unknown taxon"]

```

Convert to taxmap format for plotting:

```{r}
obj <- parse_tax_data(combined_class,
                      datasets = list(original_names = names(ref_seqs)),
                      mappings = c("{{index}}" = "{{index}}"))
```

### Filter out ambigous taxa and atrifacts

This will make the tree cleaner and easier to interpret.

```{r}
cleaned_obj <- obj %>%
  filter_taxa(is_leaf, taxon_names %in% c("Glomus", "Phytophthora"), invert = TRUE) %>%
  filter_taxa(grepl(taxon_names, pattern = 'unclassified', ignore.case = TRUE), invert = TRUE) %>%
  filter_taxa(grepl(taxon_names, pattern = 'endosymbiont', ignore.case = TRUE), invert = TRUE) %>%
  filter_ambiguous_taxa() %>%
  filter_taxa(grepl(taxon_names, pattern =  "^[a-zA-Z ]+$"))

cleaned_obj$data$tax_data <- tibble(taxon_id = names(cleaned_obj$data$tax_data), 
                                    input = cleaned_obj$data$tax_data)
```


### Plot what was amplified

I will find which portions of the taxonomy have all of their leaves amplified ...

```{r}
cleaned_obj$mutate_obs("is_amplified",
               unlist(leaves_apply(cleaned_obj,
                                    function(x) length(x) > 0 && all(x %in% names(amp_seqs)),
                                    value = "original_names")))
amplified_leafs <- cleaned_obj$data$original_names[cleaned_obj$data$original_names %in% names(amp_seqs)]
cleaned_obj$data$is_amplified[names(amplified_leafs)] <- TRUE
```

... and plot those taxa in green:

```{r}
set.seed(5)
cleaned_obj %>% 
  filter_taxa(!is_internode) %>%
  remove_redundant_names() %>%
  heat_tree(node_color = ifelse(is_amplified, "green", "grey"),
            node_color_axis_label = "PID to closest other species",
            node_size = n_obs, 
            node_size_range = c(0.005, 0.025),
            node_label_size_range = c(0.012, 0.018),
            # node_label_size_trans = "log10",
            node_size_axis_label = "Number of sequences",
            node_label = Hmisc::capitalize(taxon_names),
            layout = "da", initial_layout = "re",
            background_color = '#FFFFFF',
            # overlap_avoidance = 2,
            output_file = file.path("results", "rps10_simulated_pcr.pdf"))
```


## Which primers amplify which oomyctes

A reviewer thought it would be useful to know which of the variations of the primer amplified which organisms, with the hope that researchers could use subsets of the primers to target specific groups.

### Extract all oomycete from reference databases


```{r}
rps_ref <- read_fasta(file.path('intermediate_data', 'reference_databases', 'rps10_reference_db.fa'))
rps_ref <- rps_ref[grepl(names(rps_ref), pattern = 'Oomycetes')]
```

Remove spaces from sequences (not sure why there would be spaces).

```{r}
rps_ref <- trimws(gsub(rps_ref, pattern = ' ', replacement = ''))
```

Replace names with just species names

```{r}
names(rps_ref) <- str_match(names(rps_ref), ';([a-zA-Z0-9_\\-\'". ]+);oodb_.+;$')[,2]
```

Replace underscores with spaces

```{r}
names(rps_ref) <- gsub(names(rps_ref), pattern = '_', replacement = ' ')
```

Remove species with ambiguous names or numbers

```{r}
is_ambiguous <- function(n) {
  grepl(n, pattern = ' sp$') |
    grepl(n, pattern = ' spp$') | 
    grepl(n, pattern = ' spp.') |
    grepl(n, pattern = ' sp.') |
    grepl(n, pattern = ' aff. ')  | 
    grepl(n, pattern = ' cf.') |
    grepl(n, pattern = '[0-9]+') 
}
rps_ref <- rps_ref[! is_ambiguous(names(rps_ref))]
```

Remove information below species

```{r}
names(rps_ref) <- sub(names(rps_ref), pattern = ' var\\.?.+$', replacement = '')
names(rps_ref) <- sub(names(rps_ref), pattern = ' subsp\\.?.+$', replacement = '')
```

Now the names should be cleaned up enough:

```{r}
unique(names(rps_ref))
```


### Extract predicted amplicons

Not all reference sequences will have the entire region.
Some will have the region but not the primer sites, so these must be removed.

```{r}
primer_data <- read_csv(file.path("raw_data", "primer_data.csv"))
primer_data <- filter(primer_data, primer_id != 'rps10_f-F', locus == "rps10")
forward <- primer_data$sequence[primer_data$direction == "Forward"]
reverse <- primer_data$sequence[primer_data$direction == "Reverse"]
```

The `matchProbePair` function from the `Biostrings` package can extract a region predicted to be amplified by primers, but for some reason it does not allow for ambiguity codes, even though it is a trivial change to make it do so.
Below is a modified version of the function that does allow ambiguity codes by providing the `fixed` parameter and passing it to the `matchPattern` calls.

```{r}
matchProbePair <- function(Fprobe, Rprobe, subject, algorithm="auto",
             logfile=NULL, verbose=FALSE, fixed = FALSE)
    {
        ## This won't copy the data if Fprobe and Rprobe are already DNAString objects
        F <- DNAString(Fprobe)
        R <- DNAString(Rprobe)

        ## F and R hits on the + strand
        Fp_hits <- start(matchPattern(F, subject, algorithm=algorithm, fixed = fixed))
        Rp_hits <- start(matchPattern(R, subject, algorithm=algorithm, fixed = fixed))

        ## F and R hits on the - strand
        Fm_hits <- end(matchPattern(reverseComplement(F), subject, algorithm=algorithm, fixed = fixed))
        Rm_hits <- end(matchPattern(reverseComplement(R), subject, algorithm=algorithm, fixed = fixed))

        if (verbose) {
            cat("Fp_hits:", Fp_hits, "  Rp_hits:", Rp_hits,
                "  Fm_hits:", Fm_hits, "  Rm_hits:", Rm_hits, "\n")
        }

        matches0 <- Biostrings:::reduceProbePairMatches(c(Fp_hits, Rp_hits), c(Fm_hits, Rm_hits))
        ans <- Views(subject, start=matches0$start, end=matches0$end)

        if (!is.null(logfile)) {
            nFp <- length(Fp_hits)
            nRp <- length(Rp_hits)
            nFm <- length(Fm_hits)
            nRm <- length(Rm_hits)
            nmatches0 <- length(ans)
            ## cat("", ..., sep="\t") is a trick to get an extra tab
            cat("", nFp, nRp, nFm, nRm, nmatches0, file=logfile, sep="\t")
        }
        ans
}
```

The code below extracts predicted amplicons using primer matches, removes the primers, and aligns the trimmed amplicons to the sequences that did not have primer matches to try to find other sequences in the reference databases that do not have primer sequences but are full length matches to those that do.

```{r}
get_amplicon_chr <- function(seq) {
  map2_chr(seq@ranges@start, seq@ranges@width, function(s, w) {
    as.character(seq@subject[seq(from = s, length.out = w)])
  })
}

calc_amplicons <- function(seqs, forward, reverse, min_coverage = 0.9, ...) {
  # Get simulated amplicons using primers
  full_amps <- future_map(seqs, function(s) {
    get_amplicon_chr(matchProbePair(DNAString(s), Fprobe = forward, Rprobe = reverse))
  })
  
  # Check for multiple possible amplicons per input
  if (any(map_int(full_amps, length) > 1)) {
    stop('Some inputs have more than one amplicons.')
  }
  full_amps <- unlist(full_amps[map_int(full_amps, length) == 1])
  
  return(full_amps)
}

rps_amps <- calc_amplicons(rps_ref, 
                           forward = primer_data$sequence[primer_data$direction == 'Forward' & primer_data$locus == 'rps10'],
                           reverse = primer_data$sequence[primer_data$direction == 'Reverse' & primer_data$locus == 'rps10'])
```

The amplicon could not be extracted from many sequences, probably because it was incomplete.
Here is the proportion of reference sequences with full length amplicons extracted:

```{r}
length(rps_ref)
length(rps_amps)
```

Lets remove any sequences that are duplicates (same sequence and same species name):

```{r}
rps_amps <- rps_amps[! duplicated(paste0(names(rps_amps), rps_amps))]
length(rps_amps)
```

I noticed when debugging an issue further on in the analysis that at least one sequence is reverse complimented relative to the other sequences.
I will fix that here:

```{r}
to_rev_comp <- ! grepl(rps_amps, pattern = "^GTTGGTT")
rps_amps[to_rev_comp] <- rev_comp(rps_amps[to_rev_comp])
```

### List primer binding site sequences for each species

Cut out the primer binding region for each primer:

```{r}
forward_bindings <- substr(rps_amps, start = 0, stop = nchar(forward))
reverse_bindings <- substr(rps_amps, start = nchar(rps_amps) - nchar(reverse) + 1, stop = nchar(rps_amps)) %>%
  rev_comp()
```

check that the primer binding sites are the right sequences for the primers:

```{r}
stopifnot(all(grepl(forward_bindings, pattern = "^GTTGGTTAGAG[CT]A[AG]AAGACT$")))
stopifnot(all(grepl(reverse_bindings, pattern = "^AT[AG][CT][CT]TAGAAAGA[CT]T[CT]GAACT$")))
```

We can put this information in a table

```{r}
binding_data <- tibble(species = names(forward_bindings),
                       genus = sub(species, pattern = ' .+$', replacement = ''),
                       forward = forward_bindings,
                       reverse = reverse_bindings,
                       amplicon = rps_amps)
```

and save it for later reference: 

```{r}
binding_data %>%
  select(species, forward, reverse, amplicon) %>%
  write_csv(file = file.path('results', 'primer_binding_site_sequences.csv'))
```

An alignment of the amplicons might also be useful:

```{r}
aligned <- binding_data$amplicon %>%
  insect::char2dna() %>%
  ips::mafft(method = 'localpair', exec = '/usr/bin/mafft')
write.FASTA(aligned, file = file.path('results', 'amplicon_region_multiple_alignment.fa'))
```

### Check if primers correspond to genus

Some primers are much more common than others:

```{r}
binding_data$primer_pair <- paste(forward_bindings, "+", reverse_bindings)
table(binding_data$forward)
table(binding_data$reverse)
table(binding_data$primer_pair)
```

They don't seems to correspond the genus differences generally, but perhaps a few groups have an unique primer pair:

```{r}
(genus_bindings_forward <- t(table(binding_data[, c("genus", "forward")])))
(genus_bindings_reverse <- t(table(binding_data[, c("genus", "reverse")])))
(genus_bindings_pair <- t(table(binding_data[, c("genus", "primer_pair")])))
```

Lets save all that information in a table for future reference:

```{r}
genus_bindings_data <- rbind(
  cbind(type = 'forward', genus_bindings_forward),
  cbind(type = 'reverse', genus_bindings_reverse),
  cbind(type = 'pair', genus_bindings_pair)
)
genus_bindings_data <- cbind(primer = rownames(genus_bindings_data), genus_bindings_data)
rownames(genus_bindings_data) <- NULL
genus_bindings_data %>%
  as_tibble() %>%
  filter(type == "pair") %>%
  select(-type) %>%
  write_csv(file = file.path('results', 'genus_primer_binding_site_counts.csv'))
```



