<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Non-target amplification</title>

<script src="site_libs/jquery-1.12.4/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<script src="site_libs/htmlwidgets-1.5.2/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.16/datatables.js"></script>
<link href="site_libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.1.0.1/js/crosstalk.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Comparing ITS1 and RPS10 for Oomycete Metabacoding</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01--prepare_reference_database.html">Reference database preparation</a>
    </li>
    <li>
      <a href="02--preparation_and_quality_filtering.html">Read processing and quality filtering</a>
    </li>
    <li>
      <a href="03--abundance_matrix_preparation.html">Abundance matrix and taxonomic assignment</a>
    </li>
    <li>
      <a href="04--blast_classification.html">Alternative taxonomic assignment using BLAST</a>
    </li>
    <li>
      <a href="05--otu_clustering_threshold.html">OTU clustering threshold estimation</a>
    </li>
    <li>
      <a href="06--mock_community.html">Mock community evaluation</a>
    </li>
    <li>
      <a href="07--nontarget_amplification.html">Non-target amplification</a>
    </li>
    <li>
      <a href="08--amplicon_taxonomic_resolution.html">Taxonomic resolution</a>
    </li>
    <li>
      <a href="09--simulated_pcr_results.html">Plotting of simulated PCR results</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/grunwaldlab/rps10_barcode">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Non-target amplification</h1>

</div>


<div id="prepare" class="section level2">
<h2>Prepare</h2>
<div id="packages-used" class="section level3">
<h3>Packages used</h3>
<pre class="r"><code>library(dplyr)
library(purrr)
library(furrr)
library(tidyr)
library(readr)
library(ggplot2)
library(sessioninfo)
library(metacoder)
library(vegan)
library(viridis)
library(DT)
library(stringr)
library(qsubmitter)
library(taxize)
library(forcats)
library(ips)
library(insect)
library(ape)
library(phangorn)</code></pre>
</div>
<div id="parameters" class="section level3">
<h3>Parameters</h3>
<pre class="r"><code>minimum_read_count &lt;- 10
seed &lt;- 1
set.seed(seed)</code></pre>
</div>
</div>
<div id="detecting-non-target-amplification" class="section level2">
<h2>Detecting non-target amplification</h2>
<p>Since the reference databases will not have all the non-targets, it will be hard to detect non-target amplification. I will use blast against Genbank nt. Genbank is not curated, but it should be good enough to reliably assign a phylum.</p>
<pre class="r"><code>abundance_asv &lt;- read_csv(file.path(&#39;intermediate_data&#39;, &#39;abundance_asv.csv&#39;))</code></pre>
<pre><code>## 
## ── Column specification ────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   sequence = col_character(),
##   blast_tax = col_character(),
##   taxonomy = col_character()
## )
## ℹ Use `spec()` for the full column specifications.</code></pre>
<pre class="r"><code>abundance_asv</code></pre>
<pre><code>## # A tibble: 9,528 x 52
##    sequence blast_pid blast_cov blast_tax taxonomy    A1    A2    B1    B2    C1
##    &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 AGTCCGT…      81.5      99.3 cellular… Eukaryo…     0     0     0     0     0
##  2 TTTCCGT…      89.0      99.6 cellular… Eukaryo…     0     0     0     0     0
##  3 GAAAATC…      82.0      99.3 cellular… Eukaryo…     0     0    18 10732    17
##  4 TTTCCGT…      90.9      99.6 cellular… Eukaryo…     0     0     0     0     0
##  5 TTTCCGT…      99.7      99.0 cellular… Eukaryo…     0     0     0     0     0
##  6 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0     0     0
##  7 AGTCCGT…      81.7      99.3 cellular… Eukaryo…     0     0     0     0     0
##  8 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0     0     0
##  9 AGTCCAC…      86.8      93.9 cellular… Eukaryo…     0     0     0     0     0
## 10 GAAAATC…      85.5      99.8 cellular… Eukaryo…     0     0     0     0     0
## # … with 9,518 more rows, and 42 more variables: C2 &lt;dbl&gt;, D1 &lt;dbl&gt;, D2 &lt;dbl&gt;,
## #   E1 &lt;dbl&gt;, E2 &lt;dbl&gt;, F1 &lt;dbl&gt;, F2 &lt;dbl&gt;, G1 &lt;dbl&gt;, G2 &lt;dbl&gt;, H1 &lt;dbl&gt;,
## #   A3 &lt;dbl&gt;, A4 &lt;dbl&gt;, B3 &lt;dbl&gt;, B4 &lt;dbl&gt;, C3 &lt;dbl&gt;, C4 &lt;dbl&gt;, D3 &lt;dbl&gt;,
## #   D4 &lt;dbl&gt;, E3 &lt;dbl&gt;, E4 &lt;dbl&gt;, F3 &lt;dbl&gt;, F4 &lt;dbl&gt;, G3 &lt;dbl&gt;, G4 &lt;dbl&gt;,
## #   H3 &lt;dbl&gt;, H4 &lt;dbl&gt;, A5 &lt;dbl&gt;, A6 &lt;dbl&gt;, B5 &lt;dbl&gt;, B6 &lt;dbl&gt;, C5 &lt;dbl&gt;,
## #   C6 &lt;dbl&gt;, D5 &lt;dbl&gt;, D6 &lt;dbl&gt;, E5 &lt;dbl&gt;, E6 &lt;dbl&gt;, F5 &lt;dbl&gt;, F6 &lt;dbl&gt;,
## #   G5 &lt;dbl&gt;, G6 &lt;dbl&gt;, H5 &lt;dbl&gt;, H6 &lt;dbl&gt;</code></pre>
<pre class="r"><code>abundance_otu &lt;- read_csv(file.path(&#39;intermediate_data&#39;, &#39;abundance_otu.csv&#39;))</code></pre>
<pre><code>## 
## ── Column specification ────────────────────────────────────────────────────────
## cols(
##   .default = col_double(),
##   sequence = col_character(),
##   blast_tax = col_character(),
##   taxonomy = col_character()
## )
## ℹ Use `spec()` for the full column specifications.</code></pre>
<pre class="r"><code>abundance_otu</code></pre>
<pre><code>## # A tibble: 7,250 x 52
##    sequence blast_pid blast_cov blast_tax taxonomy    A1    A2    B1    B2    C1
##    &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 AGTCCGT…      81.5      99.3 cellular… Eukaryo…     0     0     0     0     0
##  2 TTTCCGT…      89.0      99.6 cellular… Eukaryo…     0     0     0     0     0
##  3 TTTCCGT…      90.9      99.6 cellular… Eukaryo…     0     0     0     0     0
##  4 TTTCCGT…      99.7      99.0 cellular… Eukaryo…     0     0     0     0     0
##  5 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0     0     0
##  6 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0     0     0
##  7 AGTCCAC…      86.8      93.9 cellular… Eukaryo…     0     0     0     0     0
##  8 TTTCCGT…      99.2      99.6 cellular… Eukaryo…     0     0     0     0     0
##  9 AGTCCAC…      93.7      99.3 cellular… Eukaryo…     0     0     0     0     0
## 10 TTTCCGT…     100        98.7 cellular… Eukaryo…     0     0     0     0     0
## # … with 7,240 more rows, and 42 more variables: C2 &lt;dbl&gt;, D1 &lt;dbl&gt;, D2 &lt;dbl&gt;,
## #   E1 &lt;dbl&gt;, E2 &lt;dbl&gt;, F1 &lt;dbl&gt;, F2 &lt;dbl&gt;, G1 &lt;dbl&gt;, G2 &lt;dbl&gt;, H1 &lt;dbl&gt;,
## #   A3 &lt;dbl&gt;, A4 &lt;dbl&gt;, B3 &lt;dbl&gt;, B4 &lt;dbl&gt;, C3 &lt;dbl&gt;, C4 &lt;dbl&gt;, D3 &lt;dbl&gt;,
## #   D4 &lt;dbl&gt;, E3 &lt;dbl&gt;, E4 &lt;dbl&gt;, F3 &lt;dbl&gt;, F4 &lt;dbl&gt;, G3 &lt;dbl&gt;, G4 &lt;dbl&gt;,
## #   H3 &lt;dbl&gt;, H4 &lt;dbl&gt;, A5 &lt;dbl&gt;, A6 &lt;dbl&gt;, B5 &lt;dbl&gt;, B6 &lt;dbl&gt;, C5 &lt;dbl&gt;,
## #   C6 &lt;dbl&gt;, D5 &lt;dbl&gt;, D6 &lt;dbl&gt;, E5 &lt;dbl&gt;, E6 &lt;dbl&gt;, F5 &lt;dbl&gt;, F6 &lt;dbl&gt;,
## #   G5 &lt;dbl&gt;, G6 &lt;dbl&gt;, H5 &lt;dbl&gt;, H6 &lt;dbl&gt;</code></pre>
<pre class="r"><code>metadata &lt;- read_csv(file.path(&#39;intermediate_data&#39;, &#39;metadata.csv&#39;))</code></pre>
<pre><code>## 
## ── Column specification ────────────────────────────────────────────────────────
## cols(
##   sample_id = col_character(),
##   primer_pair_id = col_character(),
##   dna_type = col_character(),
##   dna_sample_id = col_character(),
##   locus = col_character(),
##   forward = col_character(),
##   reverse = col_character(),
##   sample_type = col_character()
## )</code></pre>
<pre class="r"><code>metadata</code></pre>
<pre><code>## # A tibble: 48 x 8
##    sample_id primer_pair_id dna_type dna_sample_id locus forward reverse
##    &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;         &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  
##  1 A1        rps10_Final    mock1    mock1         rps10 rps10-F rps10-R
##  2 A2        rps10_Final    leaf     L95           rps10 rps10-F rps10-R
##  3 A3        rps10_Felipe   mock1    mock1         rps10 rps10_… rps10-R
##  4 A4        rps10_Felipe   leaf     L95           rps10 rps10_… rps10-R
##  5 A5        ITS6/7         mock1    mock1         ITS   ITS6    ITS7   
##  6 A6        ITS6/7         leaf     L95           ITS   ITS6    ITS7   
##  7 B1        rps10_Final    mock2    mock2         rps10 rps10-F rps10-R
##  8 B2        rps10_Final    WR.soil  WS3           rps10 rps10-F rps10-R
##  9 B3        rps10_Felipe   mock2    mock2         rps10 rps10_… rps10-R
## 10 B4        rps10_Felipe   WR.soil  WS3           rps10 rps10_… rps10-R
## # … with 38 more rows, and 1 more variable: sample_type &lt;chr&gt;</code></pre>
<div id="select-only-environmental-samples" class="section level3">
<h3>Select only environmental samples</h3>
<p>Non-target DNA is likely to be found the in the environmental samples, so I will restrict this analysis to those.</p>
<pre class="r"><code>metadata &lt;- filter(metadata,
                   dna_type %in% c(&quot;leaf&quot;, &quot;WR.soil&quot;, &quot;ag.soil&quot;, &quot;drip&quot;, &quot;Pan.soil&quot;),
                   primer_pair_id %in% c(&#39;rps10_Final&#39;, &#39;ITS6/7&#39;))
metadata</code></pre>
<pre><code>## # A tibble: 26 x 8
##    sample_id primer_pair_id dna_type dna_sample_id locus forward reverse
##    &lt;chr&gt;     &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;         &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;  
##  1 A2        rps10_Final    leaf     L95           rps10 rps10-F rps10-R
##  2 A6        ITS6/7         leaf     L95           ITS   ITS6    ITS7   
##  3 B2        rps10_Final    WR.soil  WS3           rps10 rps10-F rps10-R
##  4 B6        ITS6/7         WR.soil  WS3           ITS   ITS6    ITS7   
##  5 C1        rps10_Final    ag.soil  AG1           rps10 rps10-F rps10-R
##  6 C2        rps10_Final    WR.soil  WS7           rps10 rps10-F rps10-R
##  7 C5        ITS6/7         ag.soil  AG1           ITS   ITS6    ITS7   
##  8 C6        ITS6/7         WR.soil  WS7           ITS   ITS6    ITS7   
##  9 D1        rps10_Final    drip     D3            rps10 rps10-F rps10-R
## 10 D2        rps10_Final    WR.soil  WS23          rps10 rps10-F rps10-R
## # … with 16 more rows, and 1 more variable: sample_type &lt;chr&gt;</code></pre>
<pre class="r"><code>abundance_asv &lt;- select(abundance_asv, sequence:taxonomy, !!! metadata$sample_id)
abundance_asv</code></pre>
<pre><code>## # A tibble: 9,528 x 31
##    sequence blast_pid blast_cov blast_tax taxonomy    A2    A6    B2    B6    C1
##    &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 AGTCCGT…      81.5      99.3 cellular… Eukaryo…     0     0     0     0     0
##  2 TTTCCGT…      89.0      99.6 cellular… Eukaryo…     0    44     0  2988     0
##  3 GAAAATC…      82.0      99.3 cellular… Eukaryo…     0     0 10732     0    17
##  4 TTTCCGT…      90.9      99.6 cellular… Eukaryo…     0     0     0   218     0
##  5 TTTCCGT…      99.7      99.0 cellular… Eukaryo…     0     0     0     0     0
##  6 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0  1122     0
##  7 AGTCCGT…      81.7      99.3 cellular… Eukaryo…     0     0     0     0     0
##  8 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0    57     0
##  9 AGTCCAC…      86.8      93.9 cellular… Eukaryo…     0     0     0     0     0
## 10 GAAAATC…      85.5      99.8 cellular… Eukaryo…     0     0     0     0     0
## # … with 9,518 more rows, and 21 more variables: C2 &lt;dbl&gt;, C5 &lt;dbl&gt;, C6 &lt;dbl&gt;,
## #   D1 &lt;dbl&gt;, D2 &lt;dbl&gt;, D5 &lt;dbl&gt;, D6 &lt;dbl&gt;, E1 &lt;dbl&gt;, E2 &lt;dbl&gt;, E5 &lt;dbl&gt;,
## #   E6 &lt;dbl&gt;, F1 &lt;dbl&gt;, F2 &lt;dbl&gt;, F5 &lt;dbl&gt;, F6 &lt;dbl&gt;, G1 &lt;dbl&gt;, G2 &lt;dbl&gt;,
## #   G5 &lt;dbl&gt;, G6 &lt;dbl&gt;, H1 &lt;dbl&gt;, H5 &lt;dbl&gt;</code></pre>
<pre class="r"><code>abundance_otu &lt;- select(abundance_otu, sequence:taxonomy, !!! metadata$sample_id)
abundance_otu</code></pre>
<pre><code>## # A tibble: 7,250 x 31
##    sequence blast_pid blast_cov blast_tax taxonomy    A2    A6    B2    B6    C1
##    &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 AGTCCGT…      81.5      99.3 cellular… Eukaryo…     0     0     0     0     0
##  2 TTTCCGT…      89.0      99.6 cellular… Eukaryo…     0    44     0  2988     0
##  3 TTTCCGT…      90.9      99.6 cellular… Eukaryo…     0     0     0   218     0
##  4 TTTCCGT…      99.7      99.0 cellular… Eukaryo…     0     0     0     0     0
##  5 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0  1122     0
##  6 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0    57     0
##  7 AGTCCAC…      86.8      93.9 cellular… Eukaryo…     0     0     0     0     0
##  8 TTTCCGT…      99.2      99.6 cellular… Eukaryo…     0     0     0     0     0
##  9 AGTCCAC…      93.7      99.3 cellular… Eukaryo…     0     0     0     0     0
## 10 TTTCCGT…     100        98.7 cellular… Eukaryo…     0     0     0     0     0
## # … with 7,240 more rows, and 21 more variables: C2 &lt;dbl&gt;, C5 &lt;dbl&gt;, C6 &lt;dbl&gt;,
## #   D1 &lt;dbl&gt;, D2 &lt;dbl&gt;, D5 &lt;dbl&gt;, D6 &lt;dbl&gt;, E1 &lt;dbl&gt;, E2 &lt;dbl&gt;, E5 &lt;dbl&gt;,
## #   E6 &lt;dbl&gt;, F1 &lt;dbl&gt;, F2 &lt;dbl&gt;, F5 &lt;dbl&gt;, F6 &lt;dbl&gt;, G1 &lt;dbl&gt;, G2 &lt;dbl&gt;,
## #   G5 &lt;dbl&gt;, G6 &lt;dbl&gt;, H1 &lt;dbl&gt;, H5 &lt;dbl&gt;</code></pre>
</div>
<div id="filtering-out-low-abundance_asv-asvs" class="section level3">
<h3>Filtering out low abundance_asv ASVs</h3>
<p>Most low-abundance_asv ASVs would normally be filtered out during most analyses, so I will do that to.</p>
<pre class="r"><code>abundance_asv &lt;- abundance_asv[rowSums(abundance_asv[, metadata$sample_id]) &gt;= minimum_read_count, ]
abundance_asv</code></pre>
<pre><code>## # A tibble: 2,571 x 31
##    sequence blast_pid blast_cov blast_tax taxonomy    A2    A6    B2    B6    C1
##    &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 TTTCCGT…      89.0      99.6 cellular… Eukaryo…     0    44     0  2988     0
##  2 GAAAATC…      82.0      99.3 cellular… Eukaryo…     0     0 10732     0    17
##  3 TTTCCGT…      90.9      99.6 cellular… Eukaryo…     0     0     0   218     0
##  4 TTTCCGT…      99.7      99.0 cellular… Eukaryo…     0     0     0     0     0
##  5 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0  1122     0
##  6 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0    57     0
##  7 GAAAATC…      85.5      99.8 cellular… Eukaryo…     0     0     0     0     0
##  8 GAAAATC…      82.3      99.3 cellular… Eukaryo…     0     0     0     0     0
##  9 TTTCCGT…     100        98.7 cellular… Eukaryo…     0     0     0     0     0
## 10 GAAAATC…      74.3      59.2 cellular… Eukaryo…     0     0     0     0     0
## # … with 2,561 more rows, and 21 more variables: C2 &lt;dbl&gt;, C5 &lt;dbl&gt;, C6 &lt;dbl&gt;,
## #   D1 &lt;dbl&gt;, D2 &lt;dbl&gt;, D5 &lt;dbl&gt;, D6 &lt;dbl&gt;, E1 &lt;dbl&gt;, E2 &lt;dbl&gt;, E5 &lt;dbl&gt;,
## #   E6 &lt;dbl&gt;, F1 &lt;dbl&gt;, F2 &lt;dbl&gt;, F5 &lt;dbl&gt;, F6 &lt;dbl&gt;, G1 &lt;dbl&gt;, G2 &lt;dbl&gt;,
## #   G5 &lt;dbl&gt;, G6 &lt;dbl&gt;, H1 &lt;dbl&gt;, H5 &lt;dbl&gt;</code></pre>
<pre class="r"><code>abundance_otu &lt;- abundance_otu[rowSums(abundance_otu[, metadata$sample_id]) &gt;= minimum_read_count, ]
abundance_otu</code></pre>
<pre><code>## # A tibble: 1,773 x 31
##    sequence blast_pid blast_cov blast_tax taxonomy    A2    A6    B2    B6
##    &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
##  1 TTTCCGT…      89.0      99.6 cellular… Eukaryo…     0    44     0  2988
##  2 TTTCCGT…      90.9      99.6 cellular… Eukaryo…     0     0     0   218
##  3 TTTCCGT…      99.7      99.0 cellular… Eukaryo…     0     0     0     0
##  4 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0  1122
##  5 TTTCCGT…     100        99.6 cellular… Eukaryo…     0     0     0    57
##  6 TTTCCGT…     100        98.7 cellular… Eukaryo…     0     0     0     0
##  7 GAAAATC…      74.3      59.2 cellular… Eukaryo…     0     0     0     0
##  8 TTTCCGT…     100        99.4 cellular… Eukaryo…     0     0     0 86209
##  9 GAAAATC…      96.1      99.8 cellular… Eukaryo…     0     0     0     0
## 10 GAAAATC…      99.3      99.8 cellular… Eukaryo…  3542     0     0     0
## # … with 1,763 more rows, and 22 more variables: C1 &lt;dbl&gt;, C2 &lt;dbl&gt;, C5 &lt;dbl&gt;,
## #   C6 &lt;dbl&gt;, D1 &lt;dbl&gt;, D2 &lt;dbl&gt;, D5 &lt;dbl&gt;, D6 &lt;dbl&gt;, E1 &lt;dbl&gt;, E2 &lt;dbl&gt;,
## #   E5 &lt;dbl&gt;, E6 &lt;dbl&gt;, F1 &lt;dbl&gt;, F2 &lt;dbl&gt;, F5 &lt;dbl&gt;, F6 &lt;dbl&gt;, G1 &lt;dbl&gt;,
## #   G2 &lt;dbl&gt;, G5 &lt;dbl&gt;, G6 &lt;dbl&gt;, H1 &lt;dbl&gt;, H5 &lt;dbl&gt;</code></pre>
</div>
<div id="group-asvs-into-target-and-non-target" class="section level3">
<h3>Group ASVs into target and non-target</h3>
<pre class="r"><code>group_key &lt;- c(Oomycetes = &#39;Oomycetes|Oomycota&#39;, Fungi = &#39;Fungi&#39;, Bacteria = &#39;Bacteria&#39;, Plant = &#39;Viridiplantae&#39;, 
               Archaea = &#39;Archaea&#39;, Animals = &#39;Metazoa&#39;, Virus = &#39;Viruses&#39;, Protist = &#39;Jakobida|Alveolata|Euglenozoa|Amoebozoa&#39;)

groups_target_nontarget &lt;- function(abundance) {
  map_chr(abundance$blast_tax, function(x) {
    if (is.na(x)) {
      return(&#39;Unknown&#39;)
    }
    is_org &lt;- map_lgl(group_key, grepl, x = x)
    if (sum(is_org) &gt; 1) {
      stop(&#39;multiple matches&#39;)
    } else if (sum(is_org) == 1) {
      return(names(group_key)[is_org])
    } else {
      return(&#39;Other&#39;)
    }
  })
}
abundance_asv$group &lt;- groups_target_nontarget(abundance_asv)
abundance_otu$group &lt;- groups_target_nontarget(abundance_otu)</code></pre>
</div>
<div id="tally-number-of-readsasvs-per-sample-in-each-group" class="section level3">
<h3>Tally number of reads/ASVs per sample in each group</h3>
<p>I will make a table with the numbers of ASVs, reads, and the proportion of reads for each group. First the ASV counts:</p>
<pre class="r"><code>groups &lt;- unique(abundance_asv$group)

count_seqs &lt;- function(abundance, measure) {
  counts &lt;- map(metadata$sample_id, function(id) {
    map_dbl(groups, function(group) {
      is_group &lt;- abundance$group == group &amp; abundance[[id]] &gt; 0
      sum(is_group)
    })
  })
  names(counts) &lt;- metadata$sample_id
  counts &lt;- as_tibble(c(group = list(groups), 
                        measure = list(rep(measure, length(group_key))),
                        counts))
  counts
}
asv_counts &lt;- count_seqs(abundance_asv, &#39;ASVs&#39;)
otu_counts &lt;- count_seqs(abundance_otu, &#39;OTUs&#39;)</code></pre>
<p>Then read counts</p>
<pre class="r"><code>count_reads &lt;- function(abundance) {
  read_counts &lt;- map(metadata$sample_id, function(id) {
    map_dbl(groups, function(group) {
      is_group &lt;- abundance$group == group &amp; abundance[[id]] &gt; 0
      sum(abundance[[id]][is_group])
    })
  })
  names(read_counts) &lt;- metadata$sample_id
  read_counts &lt;- as_tibble(c(group = list(groups), 
                             measure = list(rep(&#39;read count&#39;, length(group_key))),
                             read_counts))
  read_counts
}
read_counts_asv &lt;- count_reads(abundance_asv)
read_counts_otu &lt;- count_reads(abundance_otu)</code></pre>
<p>And then I can convert those count to proportions:</p>
<pre class="r"><code>convert_to_prop &lt;- function(read_counts, abundance) {
  read_props &lt;- read_counts
  read_props[metadata$sample_id] &lt;- map(metadata$sample_id, function(id) {
    read_props[[id]] / sum(abundance[[id]])
  })
  read_props$measure &lt;- &#39;Reads&#39;
  read_props
}
read_props_asv &lt;- convert_to_prop(read_counts_asv, abundance_asv)
read_props_otu &lt;- convert_to_prop(read_counts_otu, abundance_otu)</code></pre>
</div>
<div id="plot-barchart-of-counts" class="section level3">
<h3>Plot barchart of counts</h3>
<p>Finally I can combine these into a single table and add some metadata</p>
<pre class="r"><code>group_abund &lt;- bind_rows(otu_counts, asv_counts, read_counts_asv, read_props_asv) %&gt;%
  gather(key = &#39;sample_id&#39;, value = &#39;abund&#39;, !!! metadata$sample_id) %&gt;%
  left_join(metadata, by = &#39;sample_id&#39;)
group_abund</code></pre>
<pre><code>## # A tibble: 832 x 11
##    group measure sample_id abund primer_pair_id dna_type dna_sample_id locus
##    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;         &lt;chr&gt;
##  1 Oomy… OTUs    A2            2 rps10_Final    leaf     L95           rps10
##  2 Fungi OTUs    A2            0 rps10_Final    leaf     L95           rps10
##  3 Other OTUs    A2            0 rps10_Final    leaf     L95           rps10
##  4 Plant OTUs    A2            0 rps10_Final    leaf     L95           rps10
##  5 Prot… OTUs    A2            0 rps10_Final    leaf     L95           rps10
##  6 Anim… OTUs    A2            0 rps10_Final    leaf     L95           rps10
##  7 Bact… OTUs    A2            0 rps10_Final    leaf     L95           rps10
##  8 Unkn… OTUs    A2            8 rps10_Final    leaf     L95           rps10
##  9 Oomy… ASVs    A2            6 rps10_Final    leaf     L95           rps10
## 10 Fungi ASVs    A2            0 rps10_Final    leaf     L95           rps10
## # … with 822 more rows, and 3 more variables: forward &lt;chr&gt;, reverse &lt;chr&gt;,
## #   sample_type &lt;chr&gt;</code></pre>
<p>I will calculate the proportion of ASVs and OTUs assigned to each group for each locus</p>
<pre class="r"><code>group_props &lt;- group_abund %&gt;%
  group_by(measure, locus) %&gt;%
  mutate(sum_abund = sum(abund)) %&gt;%
  group_by(group, .add = TRUE) %&gt;%
  summarise(prop = round(sum(abund) / sum_abund[1], digits = 4)) %&gt;%
  filter(measure %in% c(&quot;ASVs&quot;, &quot;OTUs&quot;, &quot;read count&quot;)) </code></pre>
<pre><code>## `summarise()` regrouping output by &#39;measure&#39;, &#39;locus&#39; (override with `.groups` argument)</code></pre>
<pre class="r"><code>datatable(group_props)</code></pre>
<div id="htmlwidget-2053a2454929a56e939b" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2053a2454929a56e939b">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48"],["ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","ASVs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","OTUs","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count","read count"],["ITS","ITS","ITS","ITS","ITS","ITS","ITS","ITS","rps10","rps10","rps10","rps10","rps10","rps10","rps10","rps10","ITS","ITS","ITS","ITS","ITS","ITS","ITS","ITS","rps10","rps10","rps10","rps10","rps10","rps10","rps10","rps10","ITS","ITS","ITS","ITS","ITS","ITS","ITS","ITS","rps10","rps10","rps10","rps10","rps10","rps10","rps10","rps10"],["Animals","Bacteria","Fungi","Oomycetes","Other","Plant","Protist","Unknown","Animals","Bacteria","Fungi","Oomycetes","Other","Plant","Protist","Unknown","Animals","Bacteria","Fungi","Oomycetes","Other","Plant","Protist","Unknown","Animals","Bacteria","Fungi","Oomycetes","Other","Plant","Protist","Unknown","Animals","Bacteria","Fungi","Oomycetes","Other","Plant","Protist","Unknown","Animals","Bacteria","Fungi","Oomycetes","Other","Plant","Protist","Unknown"],[0.0046,0.0021,0.8479,0.1071,0.0011,0.0307,0.0029,0.0036,0.0038,0.0171,0,0.8482,0.0095,0,0,0.1214,0.0057,0.0026,0.8966,0.0552,0.0013,0.0316,0.0031,0.0039,0.0132,0.0592,0,0.5329,0.0263,0,0,0.3684,0.0004,0,0.3631,0.6296,0.0008,0.0053,0.0005,0.0001,0.0004,0.0004,0,0.9949,0.0018,0,0,0.0023]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>measure<\/th>\n      <th>locus<\/th>\n      <th>group<\/th>\n      <th>prop<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":4},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>Check that proportions sum to 1 (there will be a bit of rounding error):</p>
<pre class="r"><code>group_props %&gt;%
  summarise(sum = sum(prop))</code></pre>
<pre><code>## `summarise()` regrouping output by &#39;measure&#39; (override with `.groups` argument)</code></pre>
<pre><code>## # A tibble: 6 x 3
## # Groups:   measure [3]
##   measure    locus   sum
##   &lt;chr&gt;      &lt;chr&gt; &lt;dbl&gt;
## 1 ASVs       ITS    1   
## 2 ASVs       rps10  1   
## 3 OTUs       ITS    1   
## 4 OTUs       rps10  1   
## 5 read count ITS    1.00
## 6 read count rps10  1.00</code></pre>
<p>Lets take a look at what proportion of oomycetes was found for each metric:</p>
<pre class="r"><code>filter(group_props, group == &quot;Oomycetes&quot;)</code></pre>
<pre><code>## # A tibble: 6 x 4
## # Groups:   measure, locus [6]
##   measure    locus group       prop
##   &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
## 1 ASVs       ITS   Oomycetes 0.107 
## 2 ASVs       rps10 Oomycetes 0.848 
## 3 OTUs       ITS   Oomycetes 0.0552
## 4 OTUs       rps10 Oomycetes 0.533 
## 5 read count ITS   Oomycetes 0.630 
## 6 read count rps10 Oomycetes 0.995</code></pre>
<p>Finally, lets plot this information:</p>
<pre class="r"><code>nontarget_plot &lt;- group_abund %&gt;% 
  filter(measure %in% c(&#39;Reads&#39;, &#39;ASVs&#39;, &#39;OTUs&#39;)) %&gt;%
  mutate(dna_type = fct_collapse(dna_type,
                                 Water = c(&#39;drip&#39;),
                                 Plant = c(&#39;leaf&#39;),
                                 Soil = c(&#39;ag.soil&#39;, &#39;Pan.soil&#39;, &#39;WR.soil&#39;)),
         measure = factor(measure, levels = c(&#39;Reads&#39;, &#39;ASVs&#39;, &#39;OTUs&#39;), ordered = TRUE),
         locus = ordered(c(rps10 = &#39;rps10&#39;, ITS = &#39;ITS1&#39;)[locus], levels = c(&#39;rps10&#39;, &#39;ITS1&#39;))) %&gt;%
  group_by(measure, locus, dna_type) %&gt;%
  mutate(abund = abund / sum(abund)) %&gt;%
  mutate(group = fct_collapse(group,
                              Oomycetes = c(&#39;Oomycetes&#39;),
                              Fungi = c(&#39;Fungi&#39;),
                              Unknown = c(&#39;Unknown&#39;),
                              Other = c(&#39;Protist&#39;, &#39;Other&#39;, &#39;Plant&#39;, &#39;Bacteria&#39;, &#39;Animals&#39;))) %&gt;%
  mutate(group = factor(group, levels = c(&quot;Unknown&quot;, &quot;Other&quot;, &quot;Fungi&quot;, &quot;Oomycetes&quot;), ordered = TRUE)) %&gt;%
  # ggplot(aes(x = locus, y = abund, fill = group)) +
  ggplot(aes(x = dna_type, y = abund, fill = group)) +
  geom_bar(stat = &#39;identity&#39;) +
  # facet_grid(measure ~ dna_type) +
  facet_grid(measure ~ locus) +
  scale_fill_viridis_d(begin = 0.8, end = 0.2) +
  labs(x = NULL, y = &#39;Proportion&#39;, fill = NULL) +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = &#39;bottom&#39;) 
ggsave(nontarget_plot, filename = &#39;non_target_amplification.pdf&#39;, path = &#39;results&#39;, width = 5, height = 5)
nontarget_plot</code></pre>
<p><img src="07--nontarget_amplification_files/figure-html/unnamed-chunk-14-1.png" width="480" /></p>
<p>Caption:</p>
<p>Figure #: Target vs non-target amplification using oomycete-specific primers for the ITS1 and RPS10 loci. Reads and ASVs are from a variety of environmental samples, grouped into soil, water, and plant tissue samples. ASV sequences were given a coarse taxonomic assignment based on BLAST searches against the NCBI nucleotide sequence database. Those assigned “Unknown” did not have a match with a E-value of at least 0.001. Sequences in “Other” include plant, animal, bacterial, and protist sequences.</p>
</div>
</div>
<div id="look-at-unknown-rps10-sequences" class="section level2">
<h2>Look at unknown rps10 sequences</h2>
<pre class="r"><code>unknown_asv_data &lt;- abundance_asv %&gt;% 
  gather(key = &#39;sample_id&#39;, value = &#39;count&#39;, !!! metadata$sample_id) %&gt;%
  left_join(metadata, by = &#39;sample_id&#39;) %&gt;%
  filter(locus == &#39;rps10&#39;, group == &#39;Unknown&#39;, count &gt; 0) %&gt;%
  select(sequence, sample_id, count, dna_type, sample_type)
datatable(unknown_asv_data)</code></pre>
<div id="htmlwidget-13f92b7856457827ab18" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-13f92b7856457827ab18">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64"],["ACTGGTATGCGTATTTTGAGTACATACTCATCGAAGCTGAACTAGACCTACCTTTTGAGCTTT","TAGCCAAAAACATGAAGAAACTGAGCAAGCCGCAGAAGAAGCCACATGTCAACCCGACGTAACTCCAAAAGCAAAATACATAGGCTGTCCTCAAACATGCACCAGCCCTCCATGTGTCTGGTTTCAAAAGGGGGCCCATTTTGGTAGCCGGTTCCCACCATTTTGTATGGACCC","TCTCCTTTGCGGGTCTGCGGAATGCGGGCGACACAAACAATCCGGTGCAATCAAAGGTGACCCCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATGGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGTATTGCAAGGCAAGCTTCTTGTGGGATTTGTCGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATCGCTGCATTTAACAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCCGCATCACCCACAATCTTGTGGATCACACTAAGGCCGAATTCG","TCTCCTTTGCGGGTCTGCGGAATGCGGGCGACACAAACAATCCGGTGCAATCAAAGGTGACCGCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATAGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGTATTGCAAGGCAAGCTTCTTGTGGGCTTTGTGGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATCGCTGCATTTAACAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCAGCATCACCCACAATGTTGTGGATCACATTAAGGCCG","TCTCCTTTGCGGGTCTGCGGAATGCGGGCGACACAAACAATCCGGTGCAATCAAAGGTGACCGCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATAGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGTATTGCAAGGCAAGCTTCTTGTGGGATTTGTCGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATCGCTGCATTTAACAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCAGCATCACCCACAATGTTGTGGATCACATTAAGGCCGAATTCG","GCACACCACCAAACTTGAACTATCCTTGGCTGCGTTCAGTGACTAATGCAACCCTTTCTACTGCCCAGGGATATAGCAGACCGGCTTTCGAACTCGGCATTGATGTG","CTTAGAAATGGTCTTTAGACTCTCTATGATCGCATAGAGACTAAAAATGGTTGGGGTGACGTCGTGAAATGGTATTAGTGACTTAGAGATGACTAGGGGAGGGTGAAGATAGGTGAAAGGGATATATATGGTAAGGTTATGGGTTGGAGGAAGGAAGGAGGTGTATGAATATATAGAAGAATAAATTATTGATATGAATGATGAAATTAATGAAAGAAAATAAAGATAACTGAAGATATATTAGGTACTTTCTGAAGAATGGATCCCAAGAGTGCAGGTGATGAAAAAAATGATGGCATAGATCCAGAAAAAAGGGATGTCATGTACATTCCACGGT","TCTCCTTTGCGGGTCTGCGGAATGCGGGCGACACAAACAATCCGGTGCAATCAAAGGTGACCCCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATGGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGCATTGCAAGGCAAGCTTCTTGTGGGATTTGTCGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATCGCTGCATTTAACAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCAGCATCACCCACAATGTTGTGGATCACATTAAGGCCGAATTCG","GCCCACGAGGACCTTAGATAACATCTCGAAAGGACATGCGTGAAGGTTCTCCTTTCTCAAA","TCTCCTTTGCGAGTCTGCGGAAGGCGGGCGACACAAACAATCCGGTGCAATAAAAGGTGACCCCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATGGCGATGGGCCGCATCGACTACGCACTGGGTGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTATGAAATTCTATTGGTATTGCAAGGCAAGCTTCTTGTGGGATTTGTCGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATTGCTGCATTTAATAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCATTCTTGCCAAGGCCTTCCGCATCACCCACAATCTTGTGGATCACACTAAGGCCAAATTCG","TCTCCTTTGCGGGTCTGCGGAATGCGGGCAACACAAACAATCCGGTGCAATCAAAGGTGACCCCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATGGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGTATGGCAAGGCAAGCTTCTTGTGGGATTTGTCGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATCGCTGCATTTAACAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCCGCATCACCCACAATCTTGTGGATCACACTAAGGCCGAATTCG","AAGGAACCGAAAGAGTGATGATTCTCCAAGCAAGGTGAAAATCTTCATAAAATGTGTGGGTGTGCATTTGCACACAGTTTTTGCACAAGAGCTTCAGGAATCACAAA","TAAGTAATTGCATCTTCAAGTTTTATTACTCTAAGTCATGTCAAGTGAAAAGGGACTTGAGTTGTGAAAAGAAAAGACTTTACTTTACTTCGATTGAAAGATTTAAGTAGTTGCACCTCCAAGTCGCACTACTTAAGTAAAGTCATTGCTCCCAAGAGAAAGACTTTACTTCATTCAAGTTGTGTCATTATGAGATACCATTGCTTAAGTAAAATAGTTTCTTCAAGTCAGTGAAGTAAAGCAAGTTGAGATCAACGCAAGCAAGGAAAAGGAGTAAAGTGAAGTAAAGAAGGTCCAAATCAACGCAGAGAGAAGGGAATAATCAAGGTTCAAATGAACGCAACAAAGGACTTAAGCAATCAGGTCTCCAATCAACGCAAAGAAAGGAAAATAAAGCAAATTGAGATCACCGATAAGAGCAGCAAGAAGAAGTGAGTTGTTCAACTACTTCAAGTAAAGTTGATTACTCCAAGTGAGAGACTTACCTCACAACTGAGTTG","TTCCCACCGCCCGCTACGAGGTGAGCGGCTCCGGCGTCGCCGCATTCATCGACTACCAAATCGTCACCGGTCAGGTGCGTGCGGTGAACGCGCAACTGCCGTGGTCGACGG","GTGAGCAGGAATCCATGCTGCTTGCCGCTGAGCGTCCCTTCGCCCACGATTTGTCCCGACGCATTGATGCCGTTGGCCGACTGCAGCACCCAACCTGATGCAGCGGAAATGAGGG","TTCGACATTTTCCAGGAGCGGGCAAGCAATCAGGTCCTGCGGCTGAATGCGTTCATGAGCCTTGCCATGTTCGACATCGACAAGATCCGCATCAAATACCGGCCGAATGGATTTCTCTGCAGCGTCCCCGGACTTAGCGCGGCGTTGAGCCTGTTCGGCAAATGCGCGCCGGTGAAGACATCAAATCCGCTGGGCTAGAATCGCCGCAGGTGGTTCTATTTGGTCCACCAAAGTTGGATTGCCATGTCGAACTTTGGTCATATCACCAATCGGAACACGGCTCCCGTTGCTGCAGCATCGATCGGATAGCCTGTTC","GCCGTAAACGCCGGGGATTGACAGTGTCCCGCCCTTGCGTACTGCCTGAATTGCCTGACGAAGCACGCTTAGCCGGTCCGTTTCCAGCATCAGGGCGGTCTTAAGCTTGTCATAGAAGCCCTGCAGTTCGGAAGAATGGGCTTCCAACCCGACTGCATCGATACACGCGTCTGGTCCGACGCCCCCGGTTAGGTCTTTGATTGCGGTCAGCACGCTCACGTCTTCTTCTGAGTAGTCGACGACGATCGCCCCCAAACTGCGCGCC","TCGGAGGATTTGGATTTCACTCCGTGGTGGTCGGGACATCGGGTTCGGGCAAGTCGGAGTTCTTCCTCTCATTGTGTAATGGGATTGCGCTGACGCATTCGCCGGAAACGTTCATCGTCATTTTCGTCGACATGA","TGTCCGGATAGGCGACCCCGAATGCAAAGCCGGTCGCCAAAAAAATTGCGAGATGCTCGAGGTTGTGTGAGGCTTCGGTTATCGGTCGATAGGAGGGGGGAACTAGGGACAGCACGACGATTGCGAGAGCAAGTAGCCACGCCGAGGCCCGACAGATTTTTTTCAGCCTAGGCCCCATCAAAAGGATCATTCGGCCTCACCAAGGTTCTAA","ACGTGATTGTCGAGTTAAATCTCTGTCGACCCACAAAGAAGGAGCTGATTGTATGCTGACGCTACCCGGGC","CGAACGGTCGCGAAACCTTGGGATGGTTTATAACGGCGATTCTTTGCGGTGGCACCTCCCGGAACGGGCAGTGTTCCTCGGTCGGTCGCCACGGCTGCGAGTATCTGTGCCTCGACGCGTGCGATCGCCGCCGTGTTATCCGAGGCCCGATAGATGTCCCGCAGTTTATTGAGTGTCGCCACCACGTTCTCGTGATGCTCGCCGTACGCCAGCCGCTCGATGTTCAAAATGCGC","TCCTTGGGATTTACCGCATGCAGCGGAGGCGACTGCGGCTGTTCCACCGCAGGGAGTGCCACCACCAACACGGACTGCGGGCTGTGAAACGAAGGGGCCGTATTCAGCGTC","TTGGCGGGATTTGGCAAGCGGATCGACGATCAAAAACAATTGGCCGAGGTGTATGCCAAATGGGCGGATGTCTTGGCGGCCAAGCAGCGGGCGGTCGTCCACCGAATGCTGACGGGACTGGCCATTGTTTTGGGAATTTTGCTGGTCGGAATCTTTTTCGATAGCTGGCTCGAACGGCTCCTCGGCAAAACGTCCCTGGATCGCAGGCAGGTCGAGACGATGCGCACGGTAACGCGCGTGGC","CTATCGCTTTCGACAACGAGAGTTCTCTTTTTGATTCAGCTGGATAGGGATCCATTCCTCCCTTCTGTAGG","TTCCCACCGCCCGCTACGAGGTGAGCGGCTCCGGCGTCGCCGCATTCATCGACTACCAAATCGTCACCGGTCAGGTGCGTGCGGTGAACGCGCAACTGCCGTGGTCGACGG","GCCGTAAACGCCGGGGATTGACAGTGTCCCGCCCTTGCGTACTGCCTGAATTGCCTGACGAAGCACGCTTAGCCGGTCCGTTTCCAGCATCAGGGCGGTCTTAAGCTTGTCATAGAAGCCCTGCAGTTCGGAAGAATGGGCTTCCAACCCGACTGCATCGATACACGCGTCTGGTCCGACGCCCCCGGTTAGGTCTTTGATTGCGGTCAGCACGCTCACGTCTTCTTCTGAGTAGTCGACGACGATCGCCCCCAAACTGCGCGCC","TCGGAGGATTTGGATTTCACTCCGTGGTGGTCGGGACATCGGGTTCGGGCAAGTCGGAGTTCTTCCTCTCATTGTGTAATGGGATTGCGCTGACGCATTCGCCGGAAACGTTCATCGTCATTTTCGTCGACATGA","TTCCCACCGCCCGCTACGAGGTGAGCGGCTCCGGCATCGCCGCATTCATCTACTACCAAACCGTCACCGGTCAGGTGCGTGCGGTGAACGCGCAACTGCCGTGGTCGACGG","TCCTTCGATATTTCGAGAATTTCCCGACCCCAACCGGCCTGCAACGGGTCGCCTTCGAGATTTACGCCGAAGCCGAACGGCTGTACGGCCGCACCGGCGCCGTGAAGTTCTGCCGGCTCAGTGTCTACACCAGGCAGTTGATGCCGGTTG","TCCTTGGGATTTACCGCATGCAGCGGAGGCGACTGCGGCTGTTCCACCGCAGGGAGTGCCACCACCAACACGGACTGCGGGCTGTGAAACGAAGGGGCCGTATTCAGCGTC","CATAATCGTTCGGTCGCGGGTTCGAGTCCCGCTCCTTCCACCAAGTAAGTAATGCAGGATTGGCATATTGGTTGTGTCCTAGCCTTCCAAGCTAGTCAAAGGAGTTCGATTCTCCTATCCTGCTCCAGTAATAAAAAGTCATGTGTCGTTGAATAGTAGATTAGTCAACTACTGCAAATAGAAGATAAGGTAAACTTAAAAACCCCCCCTTCTACATATGATAAACGCCCAGATGGTGGAATGGTATACACGATGCACTTAGAATGCATTGCCGAAAGGATTGAG","CTTGGACAGAAATTTATGTTTCTCGATCCACTTATCATTTTAATAATGATGGAGCTTTTATCGCTTTCGAACCAACAGACTCTTCTCCTATAGAATATAGGAAATCACAATCTATTAATATCTAATGATATTATCATCTTCTTTATTGAAGATGTCCGTGCCTATAGCAAAACTGGATAACGCATCAGATTACGAATCTGAAGATTGAA","TCGTGTAGCTTTGAAAGTCATTGGAGATTTAGCGTAGACAAGGATGTATTCATGGACATCTCCGATTGCGGCACGATTTTCACGTGAGTAGCGCTTCTGCCATACGTTCGTCGCGATGAAGGAATTGCGTCCGCAAACCTCGTCGAGTAGCACTCTGATGTAGGGCATTTCATAGTCGTCGATAGAAATCCATAGCGTTCCATTTGGCTCCAAGAGTCTCCAGAGCAATTCGATTCAATCTCGAATCAACGACAGCCACACGGAATGTTCAACACCGTCGTCGAAATGCTCAAAGGCTGAGCCAGTGTTGTATGGAGGATCAATATAAATGCATTTCACTTTCCCAAAGAACTCTTGCTCCAGTGCTTTGAGCGCGAGCAAGTTGTCGCCGAAAATCAAGCGATTATCGAAGATGTCATTGTCGGTCACGCGCTTCTTTGCGTGGTAGCTTTTTTCGGGATCTTCGAGCAGGATGCGAGGCTCCAGCCGTGGCCGATTCTCTTTGCCAATCCATGTC","TTCCCACCGCCCGCTACGAGGTGAGCGGCTCCGGCGTCGCCGCATTCATCGACTACCAAATCGTCACCGGTCAGGTGCGTGCGGTGAACGCGCAACTGCCGTGGTCGACGG","AAGAGGGCGATGGCGGCCGAACTGAAAACGAGCCGCTCTCAGTTGGACCGCTTGCTTGATCCCCGGAACACGGCGGTTTCGCTTGATACCTTGGCGCGGGCAGCCCACGTCCTGGGCAAGCGTTTGGTGTTGGAAATCCGAGATGCACAGCCGCAAAGATTGCGCTCTAAGAGACTCGGACAGCAGCGCACGAGAATTGCCCAGGGTACGGCCACGCGTCTCGGAAGGAGTGGCGTTTAGTTTGCCCGCGCAAGATTCTGCCACAGTTAAGTA","CATAATCGTTCGGTCGCGGGTTCGAGTCCCGCTCCTTCCACCAAGTAAGTAATGCAGGATTGGCATATTGGTTGTGTCCTAGCCTTCCAAGCTAGTCAAAGGAGTTCGATTCTCCTATCCTGCTCCAGTAATAAAAAGTCATGTGTCGTTGAATAGTAGATTAGTCAACTACTGCAAATAGAAGATAAGGTAAACTTAAAAACCCCCCCTTCTACATATGATAAACGCCCAGATGGTGGAATGGTATACACGATGCACTTAGAATGCATTGCCGAAAGGATTGAG","CTTGGACAGAAATTTATGTTTCTCGATCCACTTATCATTTTAATAATGATGGAGCTTTTATCGCTTTCGAACCAACAGACTCTTCTCCTATAGAATATAGGAAATCACAATCTATTAATATCTAATGATATTATCATCTTCTTTATTGAAGATGTCCGTGCCTATAGCAAAACTGGATAACGCATCAGATTACGAATCTGAAGATTGAA","CATAATCTGTTGGTTCGGGGTTCGAATCCCTGCGGGGGCACCATAATGTCT","GGTAGGAGAGAGGTCATGTCACGCGTCGTACTCTCCGTTGTAGTCGCAGCTGTCGCGGGGGTTTCACTCACCGGCTGTTCCTCTGCGCTGTCGTCCAACTCCTCGTCGGAACAACTCCAGGCGTTGA","GGGCTCCTCCTCCCGCGCCAGGGCATCTTCGTCAGCGCGATCGAGAACCTAGGCCGCCAGCTGATCCTCGTCAACTTTGACAGCGCCGGCCTGGAGTATTTGTTTCCCAATGAAATCTTGGCGCAGAGTCCCGGCGTTAAAGAGAGCGTTGCTGGTTCTTCCTGATAGGCTTAGGCCTAGCAGTGAGCCGCTTTTTGTTTGTTGCCTCTTGGATCAGTCGGCGTTGACTTTGACTCTTAATCAGCGGGTCCGA","TGGCGAAGCTCGTAGGTTGCTCAGATGTAAACGGTCCTGTAGGTTGTCCTCAGATGTAATCAGTCCTGGGGAATCTTGTAGGTTATCCTTGGACGTAATCAGCCCTGATGTGGTTCTATGCTAGAAATGGCCAAAGTGA","CAGATCTATATGTGGAGCTTGACACTGATGCCAAGTGTGATGTAGAAGGAGTTGGGACAATGAGGTTCT","CCAACCCGCTGACTCTGTCAAGTGCCCTGGTTTTGTTCCAGGGAATGTAATCCTTTTCCAACTGTGGAAACATGGTTTTGGAAGTTTACCATTAAGATATCTTGGGTGAGACTGTGCCTAGGATGTAGGTGTTGCCAACTGAGCTTGCTCTTTCTGGCCAATATCTTTGGGTGTTTTGTTGGTATTTGTTGA","GGGCAGGTCCGGGCAGGTTTGGGTAGGGCACCAACAGCCTAGGCCGGGTAAAACAAGTCATAGGCAGGCATAGAAGAATTCCGGGC","TGAAGCAACATCTTGATTTGTATTTCTCAACCCCAGTTGTTCCTTGCTTCACTTCTCTATTGATTTGACCATTCTTGTATGATCATGCAAACTTTAAGAAT","AGAAGAACTCAGTTCGGAGTAGGGGATATAGTCCTTTTGTGGGATAAGCATAAAGAAAAACCTGGAAATCATGGAAAGATGGACAGATTTTGGATGGGCCCCTATCGGATTGGAAGGATTGCAGGGAAAGGATCGTTTTGGTTAGAAATGCTAGATGTAGATGAGTTAGAATTTCCGGTAAATGGGTTGTTACTTAAACACTACTTTCCACCCAATGATGAGGTGACTTAGCGAGAGTTTTCCTTATTTAGTTTAGAGGAGGAATCTTGAGTAAACTCTTTTGAGCAGTAGGTTGAGTCTTTAGTGTTGCATGGTCCTGAGTCTTTGAGTCTTTACGGTTTT","AAGAGATGTTATTTTTTCAACTTTCTTGTAAGTGAAATTTGAAAGAGAATTGAGGTACTCGTTTGAGAGATAACCATGCATACAAACTTTCAAGGG","GTCGGGCTGCACTCCAAGCCATTGAAGCAGCAAAAACATGGTCAAAATCCTATTACAACTCCCATTTCCAACCTCGCACCTTCAGTGAAGGTGATATGGTTCTAGTTTATGACCAGCCCAATGATAAGTTGGGAAAAGGCA","TGGCGTTATATATTTATGTGGATAAAAGACTGATAGAAACATAAAGATACTCTGCTTCTTATATAAGAGTTAGCATTGATTCATCGACGCTGGTTCATCGTGTTCAAAGTTGGTATGGGTTGTCAAAAATTAAAATATGTCTACAAAGCATGAGGTTGTCCAAATGTTGATCATCTAGACAGTACTGCCCTCGTCTATGACAAGCCGATTTGGAGC","CTACGAATGCCTTAAAAACTGCTATATTAAAACCAGAAAATTTCTGAAGGAATTCCAACCATCCATATCTCGACACAATCTCTCGACATTCTGGAGTAGCATACAGGTATTTGTCACCAATA","AGAAGTGCTAAGTTTGGAAATGGGGATGTAGTCTGTCTGTGGGATAAATATAAAGAGAAACCAGGAAATCATGGAAAGTTGGATAAATTTTGGATGGGCCCTTACAGGGTAGGACAGAGTTCATAGAGATTGGCGACTGGTTATTCCATATAGCCCCTGGTTGTTATATAGCCCTGTATAAAGAGTAACCCAAATTATTAGCCTAGTTCAAATAGTAGTTGGAGCCCTGGTTCAAATATAGCAGGAAAGGGATCGTTTTGGTTAGAAACTTTAGATGGAGAGGAGTTGGAATTTCCAGTAAATGAGATATTACTCAAGCAATACTTTCCATCAAGTGATGGCATGACTTAGCGAGAGTTCTCCCTGTTTAGTTTAGAGTAGGAACTTGAATAAGCTTTTTTGAGCGATAGGTTTGAGTCGTTAGTGTTGCATGGTCCTAAGTCTTTGTTTTTC","CTTAGTTAGGCATGCACATTCACCTAGATTGAGTTTAACTAACCTTGTGACTCATTGAGCACTAGTTCTAATTGAACAAGTACACCCAAAACCCTAAAATATCTTAGTTGGGGTGCTACTCAGTGAAATGGAGTGCTAATGATAGTAGAGGGTT","GTGTGAGGTTTCTGGAGTCTGTGTTTGTGTCACTTGAGAACAAAGTATGGATCTGAGGTCA","AGGAAAATAGAGACATGTTGAGATACACCAAAGATTGCTAAGGGTGAAAATCAAATGAGAGAAATACAATTTCACACCACAACATGTGC","TACTTTGAAACATAGAATGTCTTCTAGTAGCAGTGTTCCTAATTGGCCATGCATTCATTGAAGTCTTACTTAACCCCTGTAAATAACTCAGAAGTAAAAGAGGTCTGGGATACT","GGAGATCCTTGGGAAGGCGAAGCTCTACATCGACGACAAGCCCGTGGCGGAAGCTCGGATACGGACGCAGAGCGGCCATTACGCCCTTTGTGGCGAGGGTCTATGCATTGGCTACGACGGTGGCGACGTGGTGTCGTCCGCGTACACGACGAAATTCGCGTTCTCAGAGGGGCGTATCATCAAGGTAGTATTCGACGTTGAGGACGACGCCTACCTGGATGTCGAGAGGAAGATGGCGGCTGCGCTGGCACGTGACTGAAGACCGCCTTGTGCCGTAAACGCCGCAAGTATTTGAATTAACAAATACGACGAGCGATCAAATCTCCCCACGAATCAGGGGGTCGGG","ACTGGTATGCGTATTTTGAGTACATACTCATCGAAGCTGAACTAGACCTACCTTTTGAGCTTT","ACCGAGCTCCATCACCCTGGCAGCCACAGATGAGTTCCGCCGGTTCTAAGGCTGCCATCCAGGCGGCAAGGGAAGGAGCTCATGTCAATATCTGGCATCCTGATGAGACTGAACAAGGGTATCGAGCAGCGGGCCAAGCCATGCGGAATACAGGCCTGAGTCCCCACGTCGAGCGAGGAGTCACAGAGGAAGGGAGCCGCCGAGCAATGCAGGCAGCCACTGGTGCCATG","TCTCCTTTGCGGGTCTGCGGAATGCGGGCGACACAAACAATCCGGTGCAATCAAAGGTGACCGCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATAGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGTATTGCAAGGCAAGCTTCTTGTGGGATTTGTCGACACTAACAATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTCGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACGTCGGCCATCAACATGCAGTGGCCATCGCTGCATTTAACAGCCAGTTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCAGCATCACCCACAATGTTGTGGATCACATTAAGGCCG","TCTCTTTTGCGGGTCTGCGGAATGCGGGCGACACAAACAATCCGGTGCAATCAAAGGTGACCCCTGCCTTTGTTCAAGAGTTTGGAGGCCTCAACACTCTGGGAATGGCGATGGGCCGCATCGACTACGCAGTGGGCGGACTAAATCCTCCTCACGCTCACCCGAGAGCCTCTGAAATTCTATTGGTATTGCAAGGCAAGCTTCTTTTGGGATTTTTCGACACTAACGATACCTTGTTCAGCAAAATATTGGAGAAGGGAGACGTGTTTGTATTTCCCAAGGCCCTGCCCCATTTCCAAGAGAACATCGGCCATCAACATGCAGTGGCCATTGCTGCATTTAACAGCCAATTTCCCGGTATTCTCACCATCGCCAATTCTCTCTTTGCTGCAAATCCTCCCATCCCCGATTCCGTTCTCGCCAAGGCCTTCCGCATCACCCACAATCTTGTGGATCACACTAAGGCCGAATTCG","AAAAGGTAAATTACGGGCCCGTGGACCTGAATTTTTATCCCAATCTGGTTGCCACGTATCCCAATCAAAAGAGGGACCCATTTTTAAGTTGGTTTACAACCTATTTGTACATAATCGATGTAAAACTTCAAAAGGGCATATCTTTCGACTCAGGGCTCGAGAG","CGCAAGAACTTACCGCAGACTCTATCCACTTGGTATGGCCCATATACTATTGCGAAGAGGGTTGGATATAAAGCATTTGAACTTAATCAACCTCCCTATATTGGTCTTCATCCAAACTTCAATGCAGAACTCTTACAGTATTACTTTCTACCTTTGTTGGATTCACCATATTGTTGAGACACTTTAACAAGCAGAGATTAATCTTTATGTGCAGCATTCTAACAGAAACGGATCACATGATTGACACCAAAGTCAAACAAACATGCCATCAAGAGATCCCTCTTTCAAATTGTCAAAGCAGATCAAGTACCTTCCCAATGAAAATAGTTCACATA","CACACTTTAAGAATTCCCAAAGAGTTGTATCGGGTAGAATTTCATGAAGAAATAGGAGATATGATTACTCTCCTCAGCAGTGTTTTTTGTCTTCCTTCTACCAAAATTTTTCACCCTTAGATGTACAAGTTCATTTGTATTATCCTTGATGCAAATATTTTCATTGATTGGGCAAGTATCGTTAGCAATACTTTACATAAGCAATTGATTGAGGTAAGGATGATAGGTCATTTTTACATGACTTCATATTTCATTTATGCTATAGTGAGAAGGAGGATATTTTCAGGGCTAACAACTAGAGGTGCAATGGGTCCGGGAAAGAATAGTGTGTTTCATTTTTACCCACAGCTGG","CCGATGCAAGAAATATTCAGATCACCATAACTTTTGACAGGGAGCTCGGATTCATGCAAGGTTGGACTTTTCGGAAATCTCTCGGCGTCCTCTACCCAACTAAGACTTGATATCTTTAATTGGGTGGGAACAAGTTCTTAAAATGCAAAAATATAGTTCTCCATTTTTGGGTTGTTGGGGGTCTCACCTACCGCCATGATGAGGAGTTAGATGCAGTCTTGGAGGTGGTTGGGGC"],["A2","A2","A2","A2","A2","A2","A2","A2","A2","A2","A2","A2","A2","B2","B2","B2","B2","B2","B2","B2","B2","B2","B2","C1","C2","C2","C2","C2","C2","C2","D1","D1","D1","D2","D2","E1","E1","E2","E2","F2","G1","G1","G1","G1","G1","G1","G1","G1","G1","G1","G1","G1","G1","G1","G1","G2","H1","H1","H1","H1","H1","H1","H1","H1"],[206,162,138,62,55,54,46,46,30,27,23,15,12,168,220,171,85,19,19,16,12,2,12,24,290,4,8,21,16,10,401,5,14,36,18,471,52,20,20,14,596,151,76,43,39,33,32,19,18,17,16,13,12,11,11,26,30,112,49,47,41,26,15,10],["leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","ag.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","WR.soil","drip","drip","drip","WR.soil","WR.soil","drip","drip","Pan.soil","Pan.soil","Pan.soil","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf","Pan.soil","leaf","leaf","leaf","leaf","leaf","leaf","leaf","leaf"],["Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample","Sample"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sequence<\/th>\n      <th>sample_id<\/th>\n      <th>count<\/th>\n      <th>dna_type<\/th>\n      <th>sample_type<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":3},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>It seems a lot these are short:</p>
<pre class="r"><code>hist(nchar(unknown_asv_data$sequence))</code></pre>
<p><img src="07--nontarget_amplification_files/figure-html/unnamed-chunk-16-1.png" width="960" /></p>
<p>And somewhat low abundance:</p>
<pre class="r"><code>hist(unknown_asv_data$count)</code></pre>
<p><img src="07--nontarget_amplification_files/figure-html/unnamed-chunk-17-1.png" width="960" /></p>
<p>The really short and low abundance ones are probably errors, so lets remove those:</p>
<pre class="r"><code>unknown_asv_data &lt;- filter(unknown_asv_data, nchar(sequence) &gt;= 100, count &gt;= 100)</code></pre>
<p>I will get a random sample of oomycete reference sequences and see if they cluster together with any in particular or form separate clades</p>
<pre class="r"><code>rps10_seqs &lt;- read_fasta(file.path(&#39;intermediate_data&#39;, &#39;reference_databases&#39;, &#39;rps10_reference_db.fa&#39;))
names(rps10_seqs) &lt;- gsub(names(rps10_seqs), pattern = &#39;;oodb_[0-9]+;$&#39;, replacement = &#39;&#39;) %&gt;%
  gsub(pattern = &#39;^.+;&#39;, replacement = &#39;&#39;) %&gt;%
  gsub(pattern = &#39;_&#39;, replacement = &#39; &#39;)
ref_subsample &lt;- rps10_seqs[!duplicated(names(rps10_seqs))]
ref_subsample &lt;- ref_subsample[sample(length(ref_subsample), 10)]</code></pre>
<p>And make a multiple sequence alignment with both the reference sequences and unknown sequences</p>
<pre class="r"><code>unknown_asv_seqs &lt;- setNames(unknown_asv_data$sequence, paste0(&#39;ASV &#39;, seq_along(unknown_asv_data$sequence), &#39; (&#39;, unknown_asv_data$count, &#39;)&#39;))
aligned &lt;- c(ref_subsample, unknown_asv_seqs) %&gt;%
  char2dna() %&gt;%
  mafft(method = &#39;localpair&#39;, exec = &#39;mafft&#39;)</code></pre>
<p>It turns out that the unknown sequences are very different from eachother and the reference sequences; so much so that I could not calculate a distance matrix to make a tree.</p>
<pre class="r"><code>image(aligned)</code></pre>
<p><img src="07--nontarget_amplification_files/figure-html/unnamed-chunk-21-1.png" width="960" /></p>
<p>I manually BLASTed a few of the sequences on NCBI’s website to verify the BLAST on all of the ASVs worked properly and did not find any close matches. These seem to be some kind of errors or mispriming against unknown organisms.</p>
</div>
<div id="software-used" class="section level2">
<h2>Software used</h2>
<pre class="r"><code>sessioninfo::session_info()</code></pre>
<pre><code>## ─ Session info ───────────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 4.0.3 (2020-10-10)
##  os       Pop!_OS 20.04 LTS           
##  system   x86_64, linux-gnu           
##  ui       X11                         
##  language en_US:en                    
##  collate  en_US.UTF-8                 
##  ctype    en_US.UTF-8                 
##  tz       America/Vancouver           
##  date     2021-05-20                  
## 
## ─ Packages ───────────────────────────────────────────────────────────────────
##  package     * version    date       lib source        
##  ade4          1.7-16     2020-10-28 [1] CRAN (R 4.0.3)
##  ape         * 5.4-1      2020-08-13 [1] CRAN (R 4.0.2)
##  aphid         1.3.3      2019-05-08 [1] CRAN (R 4.0.3)
##  askpass       1.1        2019-01-13 [1] CRAN (R 4.0.2)
##  assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.0.2)
##  bold          1.1.0      2020-06-17 [1] CRAN (R 4.0.2)
##  cli           2.1.0      2020-10-12 [1] CRAN (R 4.0.3)
##  cluster       2.1.0      2019-06-19 [4] CRAN (R 4.0.0)
##  codetools     0.2-16     2018-12-24 [4] CRAN (R 4.0.0)
##  colorspace    1.4-1      2019-03-18 [1] CRAN (R 4.0.2)
##  conditionz    0.1.0      2019-04-24 [1] CRAN (R 4.0.2)
##  crayon        1.3.4      2017-09-16 [1] CRAN (R 4.0.2)
##  crosstalk     1.1.0.1    2020-03-13 [1] CRAN (R 4.0.2)
##  crul          1.0.0      2020-07-30 [1] CRAN (R 4.0.2)
##  curl          4.3        2019-12-02 [1] CRAN (R 4.0.2)
##  data.table    1.13.2     2020-10-19 [1] CRAN (R 4.0.3)
##  digest        0.6.27     2020-10-24 [1] CRAN (R 4.0.3)
##  dplyr       * 1.0.2      2020-08-18 [1] CRAN (R 4.0.2)
##  DT          * 0.16       2020-10-13 [1] CRAN (R 4.0.3)
##  ellipsis      0.3.1      2020-05-15 [1] CRAN (R 4.0.2)
##  evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.2)
##  fansi         0.4.1      2020-01-08 [1] CRAN (R 4.0.2)
##  farver        2.0.3      2020-01-16 [1] CRAN (R 4.0.2)
##  fastmatch     1.1-0      2017-01-28 [1] CRAN (R 4.0.2)
##  forcats     * 0.5.0      2020-03-01 [1] CRAN (R 4.0.2)
##  foreach       1.5.1      2020-10-15 [1] CRAN (R 4.0.3)
##  furrr       * 0.2.1      2020-10-21 [1] CRAN (R 4.0.3)
##  future      * 1.19.1     2020-09-22 [1] CRAN (R 4.0.3)
##  generics      0.1.0      2020-10-31 [1] CRAN (R 4.0.3)
##  ggplot2     * 3.3.2      2020-06-19 [1] CRAN (R 4.0.2)
##  globals       0.13.1     2020-10-11 [1] CRAN (R 4.0.3)
##  glue          1.4.2      2020-08-27 [1] CRAN (R 4.0.2)
##  gridExtra     2.3        2017-09-09 [1] CRAN (R 4.0.3)
##  gtable        0.3.0      2019-03-25 [1] CRAN (R 4.0.2)
##  hms           0.5.3      2020-01-08 [1] CRAN (R 4.0.2)
##  htmltools     0.5.1.1    2021-01-22 [1] CRAN (R 4.0.3)
##  htmlwidgets   1.5.2      2020-10-03 [1] CRAN (R 4.0.3)
##  httpcode      0.3.0      2020-04-10 [1] CRAN (R 4.0.2)
##  igraph        1.2.6      2020-10-06 [1] CRAN (R 4.0.3)
##  insect      * 1.2.0      2018-11-25 [1] CRAN (R 4.0.3)
##  ips         * 0.0.11     2019-07-04 [1] CRAN (R 4.0.3)
##  iterators     1.0.13     2020-10-15 [1] CRAN (R 4.0.3)
##  jsonlite      1.7.1      2020-09-07 [1] CRAN (R 4.0.2)
##  kmer          1.1.2      2019-05-20 [1] CRAN (R 4.0.3)
##  knitr         1.30       2020-09-22 [1] CRAN (R 4.0.2)
##  labeling      0.4.2      2020-10-20 [1] CRAN (R 4.0.3)
##  lattice     * 0.20-41    2020-04-02 [4] CRAN (R 4.0.0)
##  lifecycle     0.2.0      2020-03-06 [1] CRAN (R 4.0.2)
##  listenv       0.8.0      2019-12-05 [1] CRAN (R 4.0.3)
##  magrittr      1.5        2014-11-22 [1] CRAN (R 4.0.2)
##  MASS          7.3-53     2020-09-09 [4] CRAN (R 4.0.2)
##  Matrix        1.2-18     2019-11-27 [4] CRAN (R 4.0.0)
##  metacoder   * 0.3.4      2020-04-29 [1] CRAN (R 4.0.3)
##  mgcv          1.8-33     2020-08-27 [4] CRAN (R 4.0.2)
##  munsell       0.5.0      2018-06-12 [1] CRAN (R 4.0.2)
##  nlme          3.1-149    2020-08-23 [4] CRAN (R 4.0.2)
##  openssl       1.4.3      2020-09-18 [1] CRAN (R 4.0.2)
##  permute     * 0.9-5      2019-03-12 [1] CRAN (R 4.0.2)
##  phangorn    * 2.5.5      2019-06-19 [1] CRAN (R 4.0.2)
##  phylogram     2.1.0      2018-06-25 [1] CRAN (R 4.0.3)
##  pillar        1.4.6      2020-07-10 [1] CRAN (R 4.0.2)
##  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.0.2)
##  plyr          1.8.6      2020-03-03 [1] CRAN (R 4.0.2)
##  prettyunits   1.1.1      2020-01-24 [1] CRAN (R 4.0.2)
##  progress      1.2.2      2019-05-16 [1] CRAN (R 4.0.2)
##  purrr       * 0.3.4      2020-04-17 [1] CRAN (R 4.0.2)
##  qsubmitter  * 0.1        2020-11-11 [1] local         
##  quadprog      1.5-8      2019-11-20 [1] CRAN (R 4.0.2)
##  R6            2.5.0      2020-10-28 [1] CRAN (R 4.0.3)
##  RANN          2.6.1      2019-01-08 [1] CRAN (R 4.0.3)
##  Rcpp          1.0.5      2020-07-06 [1] CRAN (R 4.0.2)
##  readr       * 1.4.0      2020-10-05 [1] CRAN (R 4.0.3)
##  reshape       0.8.8      2018-10-23 [1] CRAN (R 4.0.2)
##  rlang         0.4.10     2020-12-30 [1] CRAN (R 4.0.3)
##  rmarkdown     2.5        2020-10-21 [1] CRAN (R 4.0.3)
##  rstudioapi    0.11       2020-02-07 [1] CRAN (R 4.0.2)
##  scales        1.1.1      2020-05-11 [1] CRAN (R 4.0.2)
##  seqinr        4.2-4      2020-10-10 [1] CRAN (R 4.0.3)
##  sessioninfo * 1.1.1      2018-11-05 [1] CRAN (R 4.0.2)
##  sharedbib     0.1.0.9003 2020-10-16 [1] local         
##  stringi       1.5.3      2020-09-09 [1] CRAN (R 4.0.2)
##  stringr     * 1.4.0      2019-02-10 [1] CRAN (R 4.0.2)
##  taxa        * 0.3.4      2020-04-29 [1] CRAN (R 4.0.3)
##  taxize      * 0.9.99     2020-10-30 [1] CRAN (R 4.0.3)
##  tibble        3.0.4      2020-10-12 [1] CRAN (R 4.0.3)
##  tidyr       * 1.1.2      2020-08-27 [1] CRAN (R 4.0.2)
##  tidyselect    1.1.0      2020-05-11 [1] CRAN (R 4.0.2)
##  utf8          1.1.4      2018-05-24 [1] CRAN (R 4.0.2)
##  uuid          0.1-4      2020-02-26 [1] CRAN (R 4.0.2)
##  vctrs         0.3.4      2020-08-29 [1] CRAN (R 4.0.2)
##  vegan       * 2.5-6      2019-09-01 [1] CRAN (R 4.0.2)
##  viridis     * 0.5.1      2018-03-29 [1] CRAN (R 4.0.3)
##  viridisLite * 0.3.0      2018-02-01 [1] CRAN (R 4.0.2)
##  withr         2.3.0      2020-09-22 [1] CRAN (R 4.0.3)
##  xfun          0.19       2020-10-30 [1] CRAN (R 4.0.3)
##  XML           3.99-0.5   2020-07-23 [1] CRAN (R 4.0.2)
##  xml2          1.3.2      2020-04-23 [1] CRAN (R 4.0.2)
##  yaml          2.2.1      2020-02-01 [1] CRAN (R 4.0.2)
##  zoo           1.8-8      2020-05-02 [1] CRAN (R 4.0.2)
## 
## [1] /home/fosterz/R/x86_64-pc-linux-gnu-library/4.0
## [2] /usr/local/lib/R/site-library
## [3] /usr/lib/R/site-library
## [4] /usr/lib/R/library</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
